### PR TITLE
Non-genesis UTxO expenditure

### PR DESCRIPTION
**Pull request by deepfire**
1. Adds `issue-utxo-expenditure` subcommand to `cardano-cli`, with `scripts/issue-utxo-expenditure.sh`
2. Factors a couple of immediately related things:
  - `withRealPBFT`
  - `signTxId`
  - `genesisUTxOTxIn`

### CODE REVIEW

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 1]
```diff
@@ -397,25 +421,44 @@ withRealPBFT CLIOps{coProtocol} pcc common action = do
...
+txSpendGenesisUTxOByronPBFT gc sk genAddr outs =
+  let txattrs    = mkAttributes ()
       tx         = UnsafeTx (pure txIn) outs txattrs
       txIn      :: CC.UTxO.TxIn
-      txIn       = handleMissingAddr $ fst <$> Map.lookup genRichAddr initialUtxo
+      txIn       = genesisUTxOTxIn gc sk genAddr
+      cheat     :: CC.UTxO.TxWitness
+      -- Below, we have to cheat to spend a genesis UTxO entry:
+      -- sign our own TxId, not that of the input Tx.
+      -- Ledger knows.
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Run.hs]: Ledger knows what?

File: cardano-node/README.md [hunk 2]
```diff
@@ -81,33 +81,31 @@ A CLI utility to support a variety of key material operations (genesis, migratio
 
 The general synopsis is as follows:
```
Comment (Jimbo4350) [inline cardano-node/README.md]: This commit should be removed and put in a separate PR.

Comment (deepfire) [reply cardano-node/README.md]: Yes, sure, that's the intent.

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 1 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Run.hs]: Ledger has a special rule to validate TXes that use genesis UTxO -- it's in on the "cheating".

Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Run.hs]: I'll clarify the comment.

File: cardano-node/README.md [hunk 3]
```diff
@@ -81,33 +81,31 @@ A CLI utility to support a variety of key material operations (genesis, migratio
...
 
 NOTE: the exact invocation command depends on the environment.  If you have only
 built `cardano-cli`, without installing it, then you have to prepend `cabal
 new-run -- ` before `cardano-cli`.  We henceforth assume that the necessary
 environment-specific adjustment has been made, so we only mention `cardano-cli`.
 
-`SYSTEMVER` is one of the supported system generations: `byron-legacy`, `real-pbft` etc.
-The full list can be seen in the output of `cardano-cli --help`, along with the list of
-supported commands.
+The `--byron-legacy` and `--real-pbft` options immediately preceding the
```
Comment (Jimbo4350) [inline cardano-node/README.md]: :+1:

File: cardano-node/README.md [hunk 4]
```diff
@@ -116,25 +114,36 @@ Available options:
 
 ## Genesis operations
 
-The canned `scripts/genesis.sh` example provides a nice set of defaults and
-illustrates available options.  Running it will produce a `./genesis.${systemStart}` directory,
-where `${systemStart}` will be chosen 15 minutes in the future.
+### Generation
 
-The directory will have the following content:
+The genesis generation operations will create a directory that contains:
```
Comment (Jimbo4350) [inline cardano-node/README.md]: :+1:

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 5]
```diff
@@ -19,6 +20,7 @@
 {-# OPTIONS_GHC -Wno-all-missed-specialisations #-}
 {-# OPTIONS_GHC -Wno-simplifiable-class-constraints #-}
 {-# OPTIONS_GHC -Wno-partial-fields #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Run.hs]: Just saw this. Please handle the other cases explicitly.

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 6]
```diff
@@ -203,6 +209,12 @@ data ClientCommand
     Address
     (NonEmpty TxOut)
     CommonCLI
+  | SpendUTxO
+    NewTxFile
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Run.hs]: These should be annotated similarly to `IssueDelegationCertificate` above.

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 5 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Run.hs]: I found a way to convince GHC that the impossible branch is indeed impossible (removed a type variable + a type equality that pinned a branch in the `ProtocolInfo` GADT), so this is now unnecessary.

Done.

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 6 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Run.hs]: @Jimbo4350, done.

**Comment by deepfire** [comment]
All suggestions addressed @Jimbo4350 -- although I think that we could have postponed the reformatting, since it incurred some avoidable conflicts that had to be dealt with.

I also don't think we've reached consensus here, since in https://github.com/input-output-hk/cardano-node/pull/161#issuecomment-526576119 you said:

> I think for now we can use stylish-haskell for imports and use our discretion for everything else

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 7]
```diff
@@ -87,8 +89,12 @@ import           Cardano.Crypto (SigningKey (..), ProtocolMagic, ProtocolMagicId
 import qualified Cardano.Crypto.Hashing as Crypto
 import qualified Cardano.Crypto.Random as Crypto
 import qualified Cardano.Crypto.Signing as Crypto
+import           Cardano.Config.Partial (PartialCardanoConfiguration (..))
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Run.hs]: Imports here also out of order.

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 8]
```diff
@@ -196,12 +202,27 @@ data ClientCommand
   | SubmitTx
     TopologyInfo
     TxFile
+    -- ^ Filepath of transaction to submit.
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Run.hs]: :100: !

File: cardano-node/src/Cardano/CLI/Run.hs [hunk 9]
```diff
@@ -356,64 +377,113 @@ runCommand CLIOps{..}
...
-      txattrs    = mkAttributes ()
-      tx         = UnsafeTx (pure txIn) outs txattrs
-      txIn      :: CC.UTxO.TxIn
-      txIn       = handleMissingAddr $ fst <$> Map.lookup genRichAddr initialUtxo
+withRealPBFT
+  :: CLIOps IO
+  -> PartialCardanoConfiguration
+  -> CommonCLI
+  -> (RunDemo (ByronBlockOrEBB ByronConfig)
+      => CardanoConfiguration
```
Comment (Jimbo4350) [inline cardano-node/src/Cardano/CLI/Run.hs]: It's difficult for me to infer what is happening here. Can you document with comments on what `withRealPBFT` is doing and perhaps wrap 
```
(RunDemo (ByronBlockOrEBB ByronConfig)
      => CardanoConfiguration
      -> Consensus.Protocol (ByronBlockOrEBB ByronConfig)
      -> IO ())
``` 
in a `newtype` and also annotate with comments?

Comment (Jimbo4350) [reply cardano-node/src/Cardano/CLI/Run.hs]: @intricate any suggestions?

File: cardano-node/README.md [hunk 10]
```diff
@@ -206,14 +206,22 @@ transaction, in its raw wire format (see GenTx for Byron transactions).
 The canned `scripts/submit-tx.sh` script will submit the supplied transaction to a testnet
 launched by `scripts/shelley-testnet*.sh` family of scripts.
 
-### Issuing genesis UTxO expenditure Tx
+### Issuing UTxO expenditure (genesis and regular)
 
-To make a transaction spending genesis UTxO, you can either use the `issue-genesis-utxo-expenditure`
-subcommand directly, or, again use a canned script that will make a transaction tailored
-for the aforementioned testnet cluster -- `scripts/issue-genesis-utxo-expenditure.sh`.
+To make a transaction spending UTxO, you can either use the:
```
Comment (Jimbo4350) [inline cardano-node/README.md]: We don't need to do this in this PR, but I think we should migrate all instructions wrt scripts to `SCRIPTS.md` and just leave a short section in the top level `README.md` pointing to this.

File: cardano-node/app/cardano-cli.hs [hunk 11]
```diff
@@ -372,6 +393,15 @@ parseAddress opt desc =
         <> help desc
       )
 
+parseTxIn :: Parser TxIn
+parseTxIn = option (uncurry TxInUtxo
```
Comment (Jimbo4350) [inline cardano-node/app/cardano-cli.hs]: This could be reformatted to:
```
parseTxIn :: Parser TxIn
parseTxIn =
  option
    ( uncurry TxInUtxo
        . Control.Arrow.first cliParseTxId
        <$> auto
      )
    ( long "txin"
        <> metavar "(TXID,INDEX)"
        <> help "Transaction input is a pair of an UTxO TxId and a zero-based output index."
      )
```

**Comment by Jimbo4350** [comment]
> All suggestions addressed @Jimbo4350 -- although I think that we could have postponed the reformatting, since:
> 
> 1. It incurred some avoidable conflicts that had to be dealt with.
> 2. It completely invalidated the work I did (separate from this PR, but temporarily residing in it) in [6f65dc0](https://github.com/input-output-hk/cardano-node/commit/6f65dc05f9f25ffe82c01eb93de74f34fb966d3a) -- that we discussed.
> 
> I also don't think we've reached consensus here, since in [#161 (comment)](https://github.com/input-output-hk/cardano-node/pull/161#issuecomment-526576119) you said:
> 
> > I think for now we can use stylish-haskell for imports and use our discretion for everything else
> 
> ..and I thought that was the consensus -- but apparently, alas, it wasn't.

@deepfire your suggested formatting makes reading diff more difficult than it needs to be. Formatting is not only about how it looks locally.

> All suggestions addressed @Jimbo4350 -- although I think that we could have postponed the reformatting, since:
> 
> 1. It incurred some avoidable conflicts that had to be dealt with.
> 2. It completely invalidated the work I did (separate from this PR, but temporarily residing in it) in [6f65dc0](https://github.com/input-output-hk/cardano-node/commit/6f65dc05f9f25ffe82c01eb93de74f34fb966d3a) -- that we discussed.
> 
> I also don't think we've reached consensus here, since in [#161 (comment)](https://github.com/input-output-hk/cardano-node/pull/161#issuecomment-526576119) you said:
> 
> > I think for now we can use stylish-haskell for imports and use our discretion for everything else
> 
> ..and I thought that was the consensus -- but apparently, alas, it wasn't.

Unfortunately the reformatting on your branch makes it very difficult to view changes on the github diff. The local view is important but I think the format in my PR is a good compromise.

File: cardano-node/app/cardano-cli.hs [hunk 12]
```diff
@@ -8,18 +8,22 @@ import           Cardano.Chain.Common
 import           Cardano.Chain.Genesis
 import           Cardano.Chain.Slotting
 import           Cardano.Chain.UTxO
-import           Cardano.CLI.Ops (decideCLIOps)
-import           Cardano.CLI.Run
-import           Cardano.Common.Protocol
+import           Cardano.Crypto ( AProtocolMagic(..)
```
[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: @Jimbo4350, frankly I found the new order of imports confusing.

We used to have roughly the following order to the module groups:

- external pure
- external impure
- external Cardano
- cardano-node local 

The https://github.com/input-output-hk/cardano-node/pull/161 has cancelled that order.

So I'm a bit lost to what is your intent here.

File: cardano-node/README.md (hunk 10 above)
Comment (deepfire) [reply cardano-node/README.md]: I think we can address that in a future PR, what do you think?

File: cardano-node/app/cardano-cli.hs (hunk 11 above)
Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: It could be, but it seems fairly arbitrary.

I mean, how does one arrive at such a change, independently?

**Comment by deepfire** [comment]
> Unfortunately the reformatting on your branch makes it very difficult to view changes on the github diff. The local view is important but I think the format in my PR is a good compromise.

I don't think it looks too bad -- does it?:

Screenshot from 2019-08-30 15-26-42

Granted, it might be a problem in a side-by-side diff, but that's not the default view, so it's probably not used very often by most people.

Reading the code in a text editor, though, is a much more frequent operation, for most people,
and the tabular layout, I think, was quite beneficial to that use case.

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 1 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Run.hs]: Removed that piece altogether.

File: cardano-node/app/cardano-cli.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-node/app/cardano-cli.hs]: I meant alphabetically, I made a mistake in #161.

[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: So will you support the restoration of the grouping I mentioned above -- with alphabetic intra-group sorting, of course?

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 9 above)
Comment (intricate) [reply cardano-node/src/Cardano/CLI/Run.hs]: I wouldn't agree with the `newtype` point but adding a quick one-liner of doc for `withRealPBFT` would be nice.

File: cardano-node/app/cardano-cli.hs (hunk 11 above)
Comment (intricate) [reply cardano-node/app/cardano-cli.hs]: I'm also not a big fan of the formatting here. I think the usage of horizontal space in this context is what bothers me. In fact, I think I would have even been okay with a minor variant of @deepfire's formatting:
```
parseTxIn :: Parser TxIn
parseTxIn = 
  option 
    (uncurry TxInUtxo 
      . Control.Arrow.first cliParseTxId
      <$> auto
    )
    ( long       "txin"
      <> metavar "(TXID,INDEX)"
      <> help    "Transaction input is a pair of an UTxO TxId and a zero-based output index."
    )
```

Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: @intricate, would you find it acceptable to have a minor variation on your option:
```
parseTxIn :: Parser TxIn
parseTxIn = 
  option 
  ( uncurry TxInUtxo 
    . Control.Arrow.first cliParseTxId
    <$> auto
  )
  ( long       "txin"
    <> metavar "(TXID,INDEX)"
    <> help    "Transaction input is a pair of an UTxO TxId and a zero-based output index."
  )
```
The difference is that this is compatible with the default way of `haskell-mode` indentation in Emacs -- whereas the previous option isn't (it's unavailable in the sequence of indentation options it provides).

File: cardano-node/app/cardano-cli.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: Done!

File: cardano-node/src/Cardano/CLI/Run.hs (hunk 9 above)
Comment (deepfire) [reply cardano-node/src/Cardano/CLI/Run.hs]: Done!

File: cardano-node/app/cardano-cli.hs (hunk 11 above)
Comment (erikd) [reply cardano-node/app/cardano-cli.hs]: Vertically aligning code leads to git diffs with semantically null changes when the length of an identifier changes and other things are brought into alignment with it.

I currently rely heavily on git history and therefore highly value clean minimal commits with good commit messages. Anything that degrades the quality of git commits should be highly discouraged. Whoever maintains this code in the future will also rely heavily on git history. We should try and keep that history as clean and as useful as possible. This matters because we hope this code and the system it runs will be around for 50 years or more.

Comment (deepfire) [reply cardano-node/app/cardano-cli.hs]: @erikd, for the record -- I completely support that noble objective.  It's just that the original comment was unclear about the goal we were aiming for.
